mm2src/mm2_main/src/lp_swap.rs

```diff
@@ -1178,7 +1179,13 @@ pub async fn swap_kick_starts(ctx: MmArc) -> Result<HashSet<String>, String> {
     let mut coins = HashSet::new();
     let swaps = try_s!(SavedSwap::load_all_my_swaps_from_db(&ctx).await);
     for swap in swaps {
-        if swap.is_finished() {
+        if swap.is_finished_and_success()
+            || (swap.is_finished()
+                && (!swap.watcher_message_sent()
+                    || swap.refunded_by_watcher()
+                    || swap.watcher_spend_or_refund_not_found()))
```

Comment:
My goal here was not to kickstart the swap, but to trigger the `load_from_db_by_uuid` function in the beginning of the kickstart to rewrite the events if the swap was finished unsuccessfully, a watcher completes the swap in the meantime, and the taker is later restarted. For example, if the swap was finished unsucessfully with these events:

```
"Started",
"Negotiated",
"TakerFeeSent",
"TakerPaymentInstructionsReceived",
"MakerPaymentReceived",
"MakerPaymentWaitConfirmStarted",
"MakerPaymentValidatedAndConfirmed",
"TakerPaymentSent",
"WatcherMessageSent",
"TakerPaymentSpent",
"MakerPaymentSpendFailed"
"Finished"
```

and a watcher completes the swap, it will still appear as a failure on the taker which can confuse the user. That's why for unsuccessfully finished swaps, I also check if it includes the "WatcherMessageSent" event. If it does, then it is checked whether a watcher has completed the swap. If it did, I append either MakerPaymentSpentByWatcher or TakerPaymentRefundedByWatcher events and save the swap as successful.

Reply:
A swap shouldn't finish unsuccessfully unless both the taker and the watcher fail the swap. So these events/case

> For example, if the swap was finished unsuccessfully with these events:
> 
> ```
> "Started",
> "Negotiated",
> "TakerFeeSent",
> "TakerPaymentInstructionsReceived",
> "MakerPaymentReceived",
> "MakerPaymentWaitConfirmStarted",
> "MakerPaymentValidatedAndConfirmed",
> "TakerPaymentSent",
> "WatcherMessageSent",
> "TakerPaymentSpent",
> "MakerPaymentSpendFailed"
> "Finished"
> ```

should never happen.

Reply:
Do you mean, if the taker attempts to spend the maker payment and fails, it should just stop the swap without appending the "Finished" event like:

```
"Started",
"Negotiated",
"TakerFeeSent",
"TakerPaymentInstructionsReceived",
"MakerPaymentReceived",
"MakerPaymentWaitConfirmStarted",
"MakerPaymentValidatedAndConfirmed",
"TakerPaymentSent",
"WatcherMessageSent",
"TakerPaymentSpent",
"MakerPaymentSpendFailed"
```
and only after restart and checking for watcher payments append if? I think that could work, but it will require some changes. I'll think about this. I should check whether a watcher message has been sent before though. If no watcher message was sent, then I can append "Finished" right after failure.

Reply:
> Do you mean, if the taker attempts to spend the maker payment and fails, it should just stop the swap without appending the "Finished" event like

No, I mean that if the taker attempts to spend the maker payment it will be before the watcher tries that and if it failed it will be for another reason so the "Finished" event is fired. If before restart the last message was `TakerPaymentSpent`, after restart we check if the maker payment is spent by watcher or not, if spent, we will bypass the `spend_maker_payment` part using the commands for restart discussed in another comment and `MakerPaymentSpendFailed` will never be fired.

Reply:
> No, I mean that if the taker attempts to spend the maker payment it will be before the watcher tries that and if it failed it will be for another reason so the `Finished` event is fired.

But then even after the watcher completes the swap, it will appear as a failure on the taker side. Shouldn't we fix that on restart? That's what I did.

Reply:
If the taker attempts to spend the maker payment before the watcher (the default route if the taker stays online) and it fails, then the watcher will fail it too, this failure will be due to a bug in the code only and it should never happen.

> But then even after the watcher completes the swap, it will appear as a failure on the taker side. Shouldn't we fix that on restart? That's what I did.

For this case, what we fix on restart is if the last event fired was `WatcherMessageSent` or `TakerPaymentSpent` and `MakerPaymentSpent` is not part of the events (it always comes after `WatcherMessageSent`/`TakerPaymentSpent`), while the taker is offline, the payment may have been spent by the watcher so we check for that to know what the next command is (should the maker continue trying to spend the payment or should we consider the swap finished successfully).

In all cases, when `TakerSwapEvent::Finished` is found in restart that should mean that the swap is finished. I hope this is clear. If not, can you please provide me with a step by step case where this will cause problems, with the order of events fired etc..

I guess there is only one case where there is a race condition between watcher and taker where exactly before this is executed by the taker
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7ad7cc1069580273f3a383032210fa60350836ad/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1738
The watcher sends the spending payment, in this case we need to check if the payment was spent to the taker's address by the watcher or not here https://github.com/KomodoPlatform/komodo-defi-framework/blob/7ad7cc1069580273f3a383032210fa60350836ad/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1750 but this is a very edge case. I thought to cover it too :)

Reply:
> If the taker attempts to spend the maker payment before the watcher (the default route if the taker stays online) and it fails, then the watcher will fail it too, this failure will be due to a bug in the code only and it should never happen.

A bug in the code is not the only reason why the taker can fail, a connection problem can also cause a swap to fail. Think of the following scenario: The taker starts the swap, sends the watcher message and completes all the steps successfully until spending the maker payment. It will have the following events:

```
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
```
Then, when it calls the `send_taker_spends_maker_payment` function, it disconnects from the internet without being able to send the transaction, and save the swap with the following events:

```
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
    "MakerPaymentSpendFailed",
    "Finished"
```

Meanwhile, the watcher receives the watcher message, sends the transaction after waiting for some duration and completes the swap successfully. Now when the taker restarts it will have the swap saved as a failure. This is one case I could reproduce on my laptop, but I think there are more cases how a taker can fail while a watcher successfully completes a swap.

Reply:
I removed the parts where finished swaps are also checked for watcher payments, but we need to handle the disconnection cases while spending the maker payment or refunding the taker payment as we discussed offline. And it will need some thought if we want the taker to fail the swap as finished only after it is sure that the watchers are also failed, because watchers are spending/refunding only after waiting for some period to give the taker priority. For spending the maker payments, watchers wait until the taker locktime. If the taker needs to be sure that watchers have failed, it needs to wait for some period after the locktime before saving the swap as failed and finished. For refunding the taker payment, watchers wait until 1.5 * taker locktime, the taker also needs to wait for some time after before finishing the swap before failing the refund. Another solution would be to save the swap as unfinished, so it can be checked on restart. There is also the case that the taker can disconnect while confirming the taker payment or waiting for the taker payment spend, attempts to refund the taker payment and fails due to disconnection, but watcher receives the taker payment and spends the maker payment. But we can also consider having a two hour long disconnection an edge case and ignore it.

Reply:
Thanks for finding out this disconnection problems and detailing them out. Since these are not problems specific to takers using watchers only, can you please open an issue for these edge cases?